README
NAME
    DBD::OraLite - PostgreSQL emulation mode for SQLite

SUMMARY
      use DBI;
      my $dbh = DBI->connect('dbi:OraLite:dbname=file');
      # The following PostgreSQL-flavoured SQL is invalid 
      # in SQLite directly, but works using OraLite
      my $sql = q[
        SELECT
          news_id, title, cat_id, cat_name, sc_id sc_name,
          to_char(news_created,'FMDD.FMMM.YYYY') AS ndate
        FROM
          news
          NATURAL JOIN x_news_cat
          NATURAL JOIN cat
          NATURAL JOIN subcat
        WHERE
          news_active = TRUE
          AND news_created > NOW() - INTERVAL '7 days'
      ];
      my $res = $dbh->selectall_arrayref($sql,{Columns=>{}});
      # From v. 0.05 with full sequence function support
      my $get_nid = "SELECT NEXTVAL('news_news_id_seq')";
      my $news_id = $dbh->selectrow_array($get_nid);

DESCRIPTION
    The module automatically and transparently transforms a broad range of
    SQL statements typical of PostgreSQL into a form suitable for use in
    SQLite. This involves both (a) parsing and filtering of the SQL; and (b)
    the addition of several PostgreSQL-compatible functions to SQLite.

    Mainly because of datatype issues, support for many PostgreSQL features
    simply cannot be provided without elaborate planning and detailed
    metadata. Since this module is intended to be usable with any SQLite3
    database, it follows that the emulation is limited in several respects.
    An overview of what works and what doesn't is given in the following
    section on PostgreSQL Compatibility.

    DBD::OraLite has support of a sort for stored procedures. This is
    described in the Extras section below. So are the few database functions
    defined by this module which are not in PostgreSQL. Finally, the Extras
    section contains a brief mention of the DBD::OraLite::MirrorOraToSQLite
    companion module.

    If you do not want SQL filtering to be turned on by default for the
    entire session, you can connect setting the connection attribute
    *FilterSQL* to a false value:

      my $dbh = DBI->connect("dbi:OraLite:dbname=$fn",
                             undef, undef, {FilterSQL=>0});

    To turn filtering off (or on) for a single statement, you can specify
    *FilterSQL* option as a statement attribute, e.g.:

      $dbh->do($sql, {FilterSQL=>0}, @bind);
      my $sth = $dbh->prepare($sql, {FilterSQL=>0});
      $res = $dbh->selectall_arrayref($sql, {FilterSQL=>0}, @bind);

    It is possible to specify user-defined pre- and postfiltering routines,
    both globally (by specifying them as attributes of the database handle)
    and locally (by specifying them as statement attributes):

      $dbh = DBI->connect("dbi:OraLite:$file",undef,undef,
                          {prefilter=>\&prefilter});
      $res = $dbh->selectall_arrayref($sql,
                                      {postfilter=>\&postfilter},
                                      @bind_values);

    The pre-/postfiltering subroutine receives the SQL as parameter and is
    expected to return the changed SQL.

STATUS OF THE MODULE
    This module was initially developed using SQLite 3.0 and PostgreSQL 7.3,
    but it should be fully compatible with newer versions of both SQLite
    (3.1 and 3.2 have been tested) and PostgreSQL (8.1 has been tested).

    Support for SELECT statements and the WHERE-conditions of DELETE and
    UPDATE statements is rather good, though still incomplete. The module
    especially focuses on NATURAL JOIN differences and commonly used,
    built-in PostgreSQL functions.

    Support for inserted/updated values in INSERT and UPDATE statements
    could use some improvement but is useable for simple things.

    There is no support for differences in DDL.

    The SQL transformations used are not based on a formal grammar but on
    applying simple regular expressions. An obvious consequence of this is
    that they may depend excessively on the author's SQL style. YMMV. (I
    would however like you to contact me if you come across some SQL
    statements which you feel should work but that don't).

    The development of this module has been driven by personal needs, and so
    is likely to be even more one-sided than the above description suggests.

POSTGRESQL COMPATIBILITY
    In this section, the PostgreSQL functions and operators supported by the
    module are enumerated.

  Regex operators
    *   The regex operators "~", "~*", "!~" and "!~*" are transformed into
        calls to the user-defined function matches(). The regex flavour
        supported is Perl, not plain vanilla POSIX, so some
        incompatibilities may arise.

    *   Note that for ease of parsing, whitespace before and after the
        operator is required for the filtering to succeed. So "col ~ 'pat'"
        works, but "col~'pat'" doesn't.

    *   "SIMILAR TO" is not supported.

    *   ILIKE is quietly changed to LIKE. LIKE in SQLite is case-insensitive
        for 7-bit characters. In future, ILIKE will probably be handled more
        elegantly, and LIKE will be redefined so as to be more like
        PostgreSQL.

  Math Functions
    *   Added: abs, cbrt, ceil, degrees, exp, floor, ln, log (1- and
        2-argument forms), mod, pi, pow, radians, sign, sqrt, trunc (1- and
        2-argument forms), acos, asin, atan, atan2, cos, cot, sin, tan.

    *   random() exists in SQLite but was redefined to conform better with
        PostgreSQL in terms of value range. setseed() was also added, but is
        not entirely compatible with PostgreSQL in the sense that setting
        the random seed does not engender the same sequence of pseudo-random
        numbers as it would in PostgreSQL.

    *   SQLite already has a handful of mathematical functions which have
        been left alone, notably round() (1- and 2-argument forms).

  String Functions
    The only string functions which are present natively in SQLite are
    substr(), lower() and upper(). These have been left alone. Added
    functions are the following:

    *   ascii, bit_length, btrim, char_length, character_length, chr,
        convert (1- and 2-arg), decode, encode, initcap, length, lpad,
        ltrim, md5, octet_length, position, pg_client_encoding (always
        'SQL_ASCII'), quote_ident, quote_literal, repeat, replace, rpad,
        rtrim, split_part, strpos, substring(string,offset,length),
        substring(string from pattern), to_ascii (assumes latin-1 input),
        to_hex, translate, trim.

    Except for convert(), where another input encoding can be specified
    explicitly, these functions all assume that the strings are in an 8-bit
    character set, preferably iso-8859-1.

    The little-used idiom "substring(string from pattern for escape)" (where
    'pattern' is not a POSIX regular expression but a SQL pattern) is not
    supported. Otherwise support for string functions is pretty complete.

  Data Type Formatting Functions
    The implementation of these functions is impeded by the sparse type
    system employed by SQLite. Workarounds are possible, however, so this
    area will probably be better covered in future.

    *   to_char(timestamp, format) is mostly supported. There is support for
        most formatting strings (all except 'Y,YYY', 'IYYY', 'IYY', 'IY',
        'I', 'J', 'TZ', and 'tz'). The FM prefix is supported for 'MM', 'DD'
        and 'HH*', but not otherwise. Other prefixes are not supported.

    *   to_char(interval, format) and to_char(number, format) are not
        currently supported. Nor are to_date(), to_timestamp() and
        to_number() (yet).

  Date/Time Functions
    Again, SQLite's intrinsically bad support for dates and intervals makes
    this area somewhat hard to cover properly. Function support is as
    follows; also note the caveats below:

    *   Supported: now, current_date, current_time, current_datetime,
        date_part (with timestamps, not intervals), date_trunc, extract
        (with timestamps, not intervals), localtime, localtimestamp,
        timeofday.

    *   Not supported: age, isfinite, overlaps.

    Versions of SQLite 3.1 and later support some of these functions, e.g.
    current_date. In these versions the built-in will be overridden.

    The module makes no distinction between time/timestamp with and without
    time zone. It is assumed that times and timestamps are either all GMT or
    all localtime; time zone information is silently discarded. This may
    change later.

    Support for calculations with dates and intervals is still very limited.
    Basically, what is supported are expressions of the form "expr +/-
    interval 'descr'" where expr reduces to a timestamp or date value.

    If a transaction is started with begin_work(), the time as represented
    by now() and friends is "frozen" in the same way as in PostgreSQL until
    commit() or rollback() are called. A transaction started by simply
    running the SQL statement "BEGIN" does not, however, trigger this
    behaviour. Nor is the time automatically "unfrozen" when an error occurs
    during a transaction; you need to catch exceptions and call rollback()
    manually.

  Sequence Manipulation Functions
    *   There is now full support for all explicit invocations of the
        sequence functions nextval(), setval(), currval() and lastval().
        Sequences are emulated using the table pglite_seq. (This works even
        with multiple connections to the same database file, some of which
        are using transactions, since SQLite transactions lock the whole
        database file, luckily eliminating any risk of two connections
        getting the same value from a nextval() call).

        Please be aware that sequences are autogenerated if they do not
        exist. Be careful to specify the appropriate sequence names or you
        will get unexpected results.

        If a sequence being autogenerated ends with '_seq' and has a name
        which seems to match an existing table + an integer column from that
        table (tablename_colname_seq), it is given an initial value based on
        the maximum value in the column in question.

        There is as yet no support for CREATE SEQUENCE statements. Use the
        autogeneration feature to create sequences.

        Implicit calls to NEXTVAL() by omitting the serial column from the
        column list in an INSERT are caught in most cases. The main
        conditions that must be fulfilled for this to work are: (1) that the
        column in question is an integer column which is the sole primary
        key on the table; and (2) that the statement is a normal INSERT with
        a column list and a VALUES clause (and not, e.g., a statement of the
        form INSERT INTO x SELECT * FROM y).

        There is as yet no interaction with the SQLite builtin
        autoincrement/last_insert_rowid() functionality in connection with
        the sequence function support.

  Aggregate Functions
    *   max(), min(), count() and sum() are already supported by SQLite and
        have been left alone. Note that the construct "count(distinct
        colname)" is not supported unless the SQLite version being used
        supports it (3.2.6 and later).

    *   avg() has been added.

    *   stddev() and variance() are not supported.

  A Note on Casting
    Casting using the construct "::datatype" is not supported in general.
    However, "::int", "::date" and "::bool" should work as expected. All
    other casts are silently discarded.

  A Note on Booleans
    This module assumes that booleans will be stored as numeric values in
    the SQLite database. SQLite interprets 0 as false and any non-zero
    numeric value as true. Accordingly, expressions such as "= TRUE" and "=
    't'" are simply removed in SELECT and DELETE statements. Likewise, "expr
    = FALSE" is turned into "NOT expr" before being passed on to SQLite.

    In INSERT and DELETE statements, TRUE and FALSE (as well as 't'::bool
    and 'f'::bool - but not 't' and 'f' by themselves) are turned into 1 and
    0.

  Current_user etc.
    The functions current_user(), session_user() and user() - with or
    without parentheses - all mean the same thing. They return the username
    of the effective uid.

  Other Functions
    The main groups of other functions (not supported by this module at all)
    are:

    *   Database/user information functions: Aside from
        current_user/session_user/user, which were mentioned above, no
        functions in this group are supported. This includes
        current_database(), current_schema(), all functions with names
        starting with 'pg_' and 'has_', obj_description and col_description.
        See
        http://www.postgresql.org/docs/current/static/functions-misc.html

    *   Array functions are not implemented - see
        http://www.postgresql.org/docs/current/static/functions-array.html

    *   Binary string (BYTEA) functions are not implemented - see
        http://www.postgresql.org/docs/current/static/functions-binarystring
        .html

    *   Geometric functions are not implemented - see
        http://www.postgresql.org/docs/current/static/functions-geometry.htm
        l

    *   Network Address Functions are not implemented - see
        http://www.postgresql.org/docs/current/static/functions-net.html

EXTRAS
  Stored Procedures
    If the active database file contains a table called pglite_functions,
    the module assumes that it will have the following structure:

      CREATE TABLE pglite_functions (
        name   TEXT,   -- name  of the function
        argnum INT,    -- number of arguments (-1 means any number)
        type   TEXT,   -- can be 'sql' or 'perl'
        sql    TEXT,   -- the body of the function
        PRIMARY KEY (name, argnum)
      );

    In the case of a SQL-type function, it can contain syntax supported
    through the module (and not directly by SQLite). The numeric arguments
    ($1-$9) customary in PostgreSQL are supported, so that in many cases
    simple functions will be directly transferrable from pg_proc in a
    PostgreSQL database.

    An instance of a SQL snippet which would work as a function body both in
    PostgreSQL and OraLite (e.g. with the function name 'full_price_descr'):

      SELECT TRIM(group_name||': '||price_description) 
        FROM price_group NATURAL JOIN price 
        WHERE price_id = $1

    As for perl-type functions, the function body is simply the text of a
    subroutine. Here is a simple example of a function body for the function
    'commify', which takes two arguments: the number to be formatted and the
    desired number of decimal places:

      sub { 
        my ($num,$dp) = @_;
        my $format = "%.${dp}f";
        $num = scalar reverse(sprintf $format, $num);
        my $rest = $1 if $num =~ s/^(\d+)\.//;
        $num =~ s/(...)/$1,/g;
        $num = "$rest.$num" if $rest;
        return scalar reverse($num);
      }

  Non-Pg Functions
    matches(), imatches():
        These functions are used behind the scenes to implement support for
        the '~' regex-matching operator and its variants. They take two
        arguments, a string and a regular expression. matches() is case
        sensitive, imatches() isn't.

    matches_safe(), imatches_safe():
        These work in the same way as matches() and imatches() except that
        metacharacters are escaped in the regex argument. They are therefore
        in many cases more suitable for user input and other untrusted
        sources.

    lower_latin1():
        Depending on platform, lower() and upper() may not transform the
        case of non-ascii characters despite a proper locale being defined
        in the environment. This functions assumes that a Latin-1 locale is
        active and returns a lower-case version of the input given this
        assumption.

    localeorder():
        DBD::SQLite does not provide access to defining SQLite collation
        functions. This is a workaround for a specific case where this
        limitation can be an issue. Given a Latin-1 encoded string, it
        returns a string of hex digits which can be ascii-sorted in the
        ordinary way. The resulting row order will be in accordance with the
        currently active locele - but only if the locale is Latin-1 based.
        The sort is case-insensitive.

    locale():
        An information function simply returning the name of the current
        locale. The module sets the locale based on the environment
        variables $ENV{LC_COLLATE}, $ENV{LC_ALL}, $ENV{LANG}, and
        $ENV{LC_CTYPE}, in that order. Currently it is not possible to use
        different locales for character type and collation, as far as the
        module is concerned.

  DBD::OraLite::MirrorOraToSQLite
    The companion module, DBD::OraLite::MirrorOraToSQLite, may be of use in
    conjunction with this module. It can be used for easily mirroring

Changes
Revision history for Perl module DBD::OraLite

0.11 Fri Nov 21 15:32:55 2008
  - Fixed bug in selectall_hashref (by Anton Berezin)

0.10 Wed Nov 19 15:02:24 2008
  - Fixed bug in sequence support (submitted by Anton Berezin)

0.09 Thu Feb 16 13:45:22 2006
  - Fixed bug in table aliasing and column disambiguation
  - Better support for grouped tables in join clause

0.08 Wed Feb 15 10:48:56 2006
   - Much improved table aliasing support

0.07 Tue Feb 14 14:33:10 2006
   - prefilter/postfilter
   - pod change

0.06 Tue Feb 14 11:50:21 2006
   - Support for implicit nextval() in INSERT statements

0.05 Tue Feb 14 00:20:01 2006
   - Support for sequences and sequence functions
   - MirrorPgToSQLite: compatibility fix for PostgreSQL 8.1

0.04 Wed Feb  1 14:25:02 2006
   - better support for iconv on Solaris (convert function)
   - slightly improved tests
   - DBD::OraLite::MirrorOraToSQLite: changed page_limit handling
   - updated README

0.03 Mon Jan 30 16:50:00 2006
   - avoid polluting DBI::db namespace
   - small documentation fixes

0.02 Mon Jan 30 13:59:29 2006
   - prepared for distribution to CPAN

0.01 Mon Jan 30 13:31:19 2006
   - original version; created by ExtUtils::ModuleMaker 0.44



lib/DBD/OraLite/MirrorOraToSQLite.pm
package DBD::OraLite::MirrorOraToSQLite;
use strict;
use DBI;
use Storable;
use File::Copy;
require Exporter;
our @ISA = qw(Exporter);
our @EXPORT_OK = qw(pg_to_sqlite);
our $VERSION = '0.05';

#### MAIN SUBROUTINE ####

sub pg_to_sqlite {
	my %opt = @_;
	my %defaults = defaults();
	for (keys %defaults) {
		next if exists $opt{$_};
		$opt{$_} = $defaults{$_};
	}
	$opt{tables} = _commasplit($opt{tables});
	$opt{views}  = _commasplit($opt{views});
	if ($opt{where}
		&& $opt{where} !~ /^\s*where\s/i 
		&& $opt{where} !~ /^\s*(?:natural\s+)?join\s/i) {
		$opt{where} = 'where '.$opt{where};
	}
	die "Incompatible options: 'append' and 'snapshot'" if $opt{append} && $opt{snapshot};
	die "Need either database handle (pg_dbh) or DSN (pg_dsn)" unless $opt{pg_dbh} || $opt{pg_dsn};
	my $disconnect = 0;
	unless ($opt{pg_dbh}) {
		$disconnect++;
		$opt{pg_dbh} = DBI->connect(@opt{ qw(pg_dsn pg_user pg_pass) },{RaiseError=>1})
		  or die "Could not connect to Oracle: $DBI::errstr";
	}
	my $fn = $opt{sqlite_file};
	die "Need both list of source tables and a SQLite file" unless @{$opt{tables}} && $fn;
	$|++ if $opt{verbose};
	my $lockfile = "$fn.lock";
	lockfile('create',$lockfile);
	if (-f "$fn.tmp") {
		warn "WARNING: Removing temp file $fn.tmp - apparently left over from a previous run\n";
		unlink "$fn.tmp" or die "ERROR: Could not remove file.tmp: $!\n";
	}
	if ($opt{append}) {
		unless (copy $fn, "$fn.tmp") {
			warn "WARNING: Could not copy $fn to $fn.tmp for appending: $! - turning --append off\n";
			$opt{append} = 0;
		}
	}
	$opt{sl_dbh} = DBI->connect("dbi:SQLite:dbname=$fn.tmp",undef,undef,{RaiseError=>1});

	my @tables = tablelist($opt{pg_dbh}, $opt{schema}, @{ $opt{tables} }); # handle regexp
	print "MIRRORING ".scalar(@tables)." table(s):\n" if $opt{verbose};
	my @views = viewlist($opt{pg_dbh}, $opt{schema}, @{ $opt{views} }) 
	  if grep { /^\/.+\/$/ } @{ $opt{views} };

	eval {
		if ($opt{snapshot}) {
			$opt{pg_dbh}->do("set session characteristics as transaction isolation level serializable");
			$opt{pg_dbh}->begin_work;
			mirror_table($_,%opt) for @tables;
			mirror_functions(%opt) if $opt{functions};
			$opt{pg_dbh}->commit;
		} else {
			mirror_table($_,%opt) for @tables;
			mirror_functions(%opt) if $opt{functions};
		}
		if (@views) {
			print "CREATING ".scalar(@views)." view(s):\n" if $opt{verbose};
			create_view($_,%opt) for @views;
		}
		print "done!\n" if $opt{verbose};
	};

	if ($@) {
		lockfile('clear',$lockfile);
		die $@;
	}

	$opt{pg_dbh}->disconnect if $disconnect;
	$opt{sl_dbh}->disconnect;

	if (-f $fn) {
		copy $fn, "$fn.bak" or warn "WARNING: Could not make backup copy of $fn: $!\n";
	}
	move "$fn.tmp", $fn or die "ERROR: Could not move temporary SQLite file $fn.tmp to $fn";
	lockfile('clear',$lockfile);
}



########## OTHER SUBROUTINES ###########

sub _commasplit {
	my $list = shift;
	return [] unless $list;
	$list = [$list] unless ref $list;
	my @new = split(/\s*,\s*/,join(',',@$list));
	return \@new;
}

sub defaults {
	my %defaults = (
					verbose     => 0,
					pg_dsn      => ($ENV{PGDATABASE} ? "dbi:Pg:dbname=$ENV{PGDATABASE}" : undef),
					pg_user     => ($ENV{PGUSER} || $ENV{USER}),
					pg_pass     => $ENV{PGPASSWORD},
					schema      => 'public',
					tables      => [],
					sqlite_file => '',
					where       => '',
					cachedir    => '/tmp/sqlite_mirror_cache',
					append      => 0,
					snapshot    => 0,
					indexes     => 0,
					views       => [],
					functions   => 0,
					page_limit  => 5000, # each page is 8K
					pg_dbh      => undef,
				   );
	$defaults{pg_dsn} ||= "dbi:Pg:dbname=$defaults{pg_user}" if $defaults{pg_user};
	return %defaults;
}

sub mirror_table {
	my ($tn,%opt) = @_;
	my $sn = $opt{schema};
	print "  - $sn.$tn\n" if $opt{verbose};
	my ($create,$colcnt) = get_schema($sn,$tn,%opt);
	my $drop = '';
	if ($opt{append}) {
		$drop = $opt{sl_dbh}->selectrow_array("select name from sqlite_master where type = 'table' and name = ?",{},$tn);
		$opt{sl_dbh}->do("drop table $tn") if $drop;
	}
	$opt{sl_dbh}->do($create);
	my $pages = $opt{pg_dbh}->selectrow_array("select relpages from pg_class where relnamespace = (select oid from pg_namespace where nspname = ?) and relname = ?",{},$sn,$tn);
	my $ins = $opt{sl_dbh}->prepare("insert into $tn values (". join(',', ("?") x $colcnt) . ")");
	if ($pages > $opt{page_limit}) {
		warn "      pagelimit ($opt{page_limit}) kicks in for $sn.$tn ($pages)\n" if $opt{verbose};
		my @pkey = $opt{pg_dbh}->primary_key(undef,$sn,$tn);
		warn "         (pkey is )".join(":",@pkey)."\n" if $opt{verbose};
		if (@pkey) {
			my $pkey_vals = $opt{pg_dbh}->selectall_arrayref("select ".join(', ', @pkey)." from $sn.$tn");
			my $sql = "select * from $sn.$tn where ".join(" and ", map {"$_ = ?"} @pkey);
			my $selh = $opt{pg_dbh}->prepare($sql);
			$opt{sl_dbh}->begin_work;
			foreach (@$pkey_vals) {
				$selh->execute(@$_);
				my $row = $selh->fetchrow_arrayref;
				$ins->execute(@$row);
			}
			$opt{sl_dbh}->commit;
			$selh->finish;
		}
		else {
			warn "*** CANNOT READ $sn.$tn ROW-BY-ROW - NO PRIMARY KEY\n*** SKIPPING TABLE!\n";
			$ins->finish;
			return;
		}
	}
	else {
		my $res = $opt{pg_dbh}->selectall_arrayref("select * from $sn.$tn $opt{where}");
		if (@$res && scalar(@{$res->[0]}) != $colcnt) {
			$ins->finish;
			die "ERROR: Bad schema for table $tn: number of columns does not match\n";
		}
		$opt{sl_dbh}->begin_work;
		$ins->execute(@$_) for @$res;
		$opt{sl_dbh}->commit;
	}
	$ins->finish;
	create_indexes($tn,%opt) if $opt{indexes};
	$opt{sl_dbh}->do("vacuum") if $drop;
}

sub mirror_functions {
	my %opt = @_;
	unless ($opt{sl_dbh}->selectrow_array("select name from sqlite_master where name = 'pglite_functions' and type = 'table'")) {
		$opt{sl_dbh}->do("CREATE TABLE pglite_functions (name text, argnum int, type text, sql text, primary key (name,argnum))");
	}
	my $langnum = $opt{pg_dbh}->selectrow_array("select oid from pg_language where lanname = 'sql'");
	my $snum = $opt{pg_dbh}->selectrow_array("select oid from pg_namespace where nspname = ?",{},$opt{schema});
	my $fun = $opt{pg_dbh}->selectall_arrayref("select proname, pronargs, prosrc from pg_proc where prolang = ? and pronamespace = ?",{},$langnum,$snum);
	print "FUNCTIONS:\n" if $opt{verbose};
	return unless ref $fun eq 'ARRAY' && @$fun;
	for (@$fun) {
		my ($name,$argnum,$sql) = @$_;
		unless ($opt{sl_dbh}->selectrow_array("select name from pglite_functions where name = ? and argnum = ?",{},$name,$argnum)) {
			print "    - $name ($argnum)\n" if $opt{verbose};
			$opt{sl_dbh}->do("insert into pglite_functions (name,argnum,type,sql) values (?,?,'sql',?)",{},$name,$argnum,$sql);
		}
	}
}

sub create_indexes {
	my ($tn,%opt) = @_;
	my $sn = $opt{schema};
	my $ixn = $opt{pg_dbh}->selectcol_arrayref("select indexdef from pg_indexes where schemaname = ? and tablename = ? ", {}, $sn,$tn);
	for (@$ixn) {
		next if /\(oid\)/;
		next if /_pkey\b/; # No need to recreate primary keys
		s/USING btree //;
		s/ ON \w+\.\"?([^\"]+)\"?/ ON $1/;
		print "      + $_\n" if $opt{verbose};
		eval {  $opt{sl_dbh}->do($_); };
		# Pg supports multiple null values in  unique columns - SQLite doesn't
		if ($@ =~ /unique/i && s/ UNIQUE / /) {
			print "      + retry: $_\n" if $opt{verbose};
			eval { $opt{sl_dbh}->do($_); };
		}
	}
}

sub create_view {
	my ($vn,%opt) = @_;
	my $sn = $opt{schema};
	my $def = $opt{pg_dbh}->selectrow_array("select definition from pg_views where schemaname = ? and viewname = ?",{},$sn,$vn);
	print "  - $sn.$vn\n" if $opt{verbose};
	$def =~ s/::\w+//g; # casting is not supported in SQLite
	if ($opt{sl_dbh}->selectrow_array("select name from sqlite_master where name = ? and type = 'view'",{},$vn)) {
		eval { $opt{sl_dbh}->do("drop view $vn") };
	}
	eval { $opt{sl_dbh}->do("create view $vn as $def"); };
	warn "    *** COULD NOT CREATE VIEW $sn.$vn *** \n" if $@;
}

sub tablelist {
	my ($pg,$sn,@pats) = @_;
	my %tables; # prevent duplicate table names
	for my $pat (@pats) {
		if ($pat =~ s/^\/(.+)\/$/$1/) {
			my $res = $pg->selectcol_arrayref("select tablename from pg_tables where lower(schemaname) = lower('$sn') and tablename ~* '$pat'");
			$tables{$_}++ for @$res;
		} else {
			$tables{$pat}++;
		}
	}
	return keys %tables;
}

sub viewlist {
	my ($pg,$sn,@pats) = @_;
	my %views; # prevent duplicate table names
	for my $pat (@pats) {
		if ($pat =~ s/^\/(.+)\/$/$1/) {
			my $res = $pg->selectcol_arrayref("select viewname from pg_views where lower(schemaname) = lower('$sn') and viewname ~* '$pat'");
			$views{$_}++ for @$res;
		} else {
			$views{$pat}++;
		}
	}
	return keys %views;
}


sub get_schema {
	my ($sn,$tn,%opt) = @_;
	# Constructing a schema definition can be rather slow,
	# so we cache the result for up to a week
	my @cached = cached_schema($sn,$tn,undef,undef,%opt);
	return @cached if @cached;
	my @cdef = col_def($sn,$tn,%opt);
	my $colcnt = scalar @cdef;
	my @pknames = $opt{pg_dbh}->primary_key(undef,$sn,$tn);
	push @cdef, "primary key (" . join(',',@pknames) . ")" if @pknames && $pknames[0] ne '';
	my $create = "create table $tn (\n  ".join(",\n  ",@cdef)."\n)\n";
	cached_schema($sn,$tn,$create,$colcnt,%opt);
	return ($create, $colcnt);
}

sub cached_schema {
	my ($sn,$tn,$creat,$cnt,%opt) = @_;
	my $database = $opt{pg_dbh}->{mbl_dbh} 
	  ? $opt{pg_dbh}->{mbl_dbh}->[0]->{Name}
	  : $opt{pg_dbh}->{Name};
	unless (-d $opt{cachedir}) {
		mkdir $opt{cachedir};
		chmod 0777, $opt{cachedir};
	}
	my $uid = (getpwuid($>))[0] || $>;
	mkdir "$opt{cachedir}/$uid" unless -d "$opt{cachedir}/$uid";
	my $fn = "$opt{cachedir}/$uid/$database.$sn.$tn";
	if ($cnt) {
		Storable::store [$creat,$cnt], $fn;
	} elsif (-f $fn && time-(stat $fn)[9]<7*24*60*60) {
		my $ret = Storable::retrieve $fn || [];
		return @$ret;
	} else {
		return ();
	}
}

sub col_def {
	my ($sn,$tn,%opt) = @_;
	my $sth = $opt{pg_dbh}->column_info(undef,$sn,$tn,undef);
	$sth->execute;
	my $res = $sth->fetchall_arrayref;
	my @ret;
	foreach my $ci (@$res) {
		my ($colnam,$typnam,$nullable) = @{$ci}[qw(3 5 10)]; #)];
		my $notnull = $nullable ? "" : " not null";
		push @ret, "$colnam $typnam$notnull";
	}
	$sth->finish;
	return @ret;
}

sub lockfile {
	my ($action,$lockfile) = @_;
	if ($action eq 'create') {
		die "ERROR: Lockfile $lockfile exists - cannot continue" if -f $lockfile;
		open LOCK, ">", "$lockfile" or die "ERROR: Could not open lockfile $lockfile: $!";
		print LOCK $$;
		close LOCK;
	} elsif ($action eq 'clear') {
		if (-f $lockfile) {
			unlink $lockfile or die "ERROR: Could not remove lockfile $lockfile: $!";
		} else {
			warn "WARNING: Lockfile $lockfile does not exist - cannot clear" unless -f $lockfile;
		}
	}
}

1;

__END__

=pod

=head1 NAME

DBD::OraLite::MirrorOraToSQLite - Mirror tables from Oracle to SQLite

=head1 SUMMARY

 use DBD::OraLite::MirrorOraToSQLite qw(pg_to_sqlite);
 pg_to_sqlite(
     sqlite_file => '/var/pg_mirror/news.sqlite',
     pg_dbh      => $dbh,
     schema      => 'news',
     tables      => [ qw(news cat img /^x_news/)],
     views       => [ 'v_newslist' ],
     indexes     => 1,
     verbose     => 1,
     snapshot    => 1,
 );

=head1 USAGE

The purpose of this module is to facilitate mirroring of tables from a
Oracle database to a SQLite file. The module has only be tested
with PostgreSQL 7.3 and SQLite 3.0-3.2. SQLite 2.x will probably not
work; as for PostgreSQL, any version after 7.2 is supposed to work. If
it doesn't, please let me know.

As seen above, options to the pg_to_sqlite() function (which is
exported on request) are passed in as a hash.  These options are
described below. The default values can be changed by overriding the
DBD::OraLite::MirrorOraToSQLite::defaults() subroutine.

=head2 Required options

Obviously, the mirroring function needs either a PosgtgreSQL database
connection or enough information to be able to connect to the database
by itself. It also needs the name of a target SQLite file, and a list
of tables to copy between the two databases.

=over 4

=item pg_dbh, pg_user, pg_pass, pg_dsn

If a database handle is specified in I<pg_dbh>, it takes
precedence. Otherwise we try to connect using I<pg_dsn>, I<pg_user>,
and I<pg_pass> (which are assigned defaults based on the environment
variables PGDATABASE, PGUSER and PGPASSWORD, if any of these is
present).

=item tables

The value of the required I<tables> option should be an arrayref of
strings or a string containing a comma-separated list of tablenames
and tablename patterns. A tablename pattern is a string or distinct
string portion delimited by forward slashes. To clarify: Suppose that
a database contains the tables news, img, img_group, cat, users,
comments, news_read_log, x_news_cat, x_news_img, and x_img_group; and
that we want to mirror news, img, cat, x_news_img and x_news_cat,
leaving the other tables alone. To achieve this, you would set the
I<tables> option to any of the following (there are of course also
other possibilities):

 (1) [qw(news img cat x_news_img x_news_cat)]
 (2) 'news, img, cat, x_news_img, x_news_cat'
 (3) [qw(news /img$/ /cat$/)]
 (4) 'news,/img$/,/cat/'

The purpose of this seemingly unneccesary flexibility in how the table
list is specified is to make the functionality of the module more
easily accessible from the command line.

Please note that the patterns between the slash delimiters are not
Perl regular expressions but rather POSIX regular expressions, used to
query the PostgreSQL system tables directly.

=item sqlite_file

This should specify the full path to a SQLite file. While the
mirroring takes place, the incoming data is not written directly to
this file, but to a file with the same name except for a '.tmp'
extension. When the operation has finished, the previous file with the
name specified (if any) is renamed with a '.bak' extension, and the
.tmp file is renamed to the requested filename. Unless you use the
I<append> option, the information previously in the file will be
totally replaced.

=back

=head2 Other options

=over 4

=item schema

This signifies the schema from which the tables on the PostgreSQL side
are to be fetched. Default: 'public'. Only one schema can be specified
at a time.

=item where

A WHERE-condition appended to the SELECT-statement used to get data
from the PostgreSQL tables.

=item views

A list of views, specified in the same manner as the list of tables
for the I<tables> option. An attempt is made to define corresponding
views on the SQLite side (though this functionality is far from
reliable).

=item indexes

A boolean option indicating whether to create indexes for the same
columns in SQLite as in PostgreSQL. Default: false. (Normally only the
primary key is created).

=item functions

A boolean indicating whether to attempt to create functions on the
SQLite side corresponding to any SQL language (NOT PL/pgSQL or other
procedural language) functions in the PostgreSQL database. This is
for use with DBD::OraLite only, since these functions are put into the
pglite_functions table. Default: false.

=item page_limit

Normally the information from the PostgreSQL tables is read into
memory in one go and transferred directly to the SQLite file. This is,
however, obviously not desireable for very large tables. If the
PostgreSQL system tables report that the page count for the table is
above the limit specified by I<page_limit>, the table is instead
transferred row-by-row. Default value: 5000; since each page normally
is 8K, this represents about 40 MB on disk and perhaps 70-100 MB of
memory usage by the Perl process. For page_limit to work, the table
must have a primary key.

NB! Do not set this limit lower than necessary: it is orders of
magnitude slower than the default "slurp into memory" mode.

=item append

If this boolean option is true, then instead of creating a new SQLite
file, the current contents of the I<sqlite_file> are added to. If a
table which is being mirrored existed previously in the file, it is
dropped and recreated, but any tables not being copied from PostgreSQL
in the current run are left alone. (This is primarily useful for
mirroring some tables in toto, and others only in part, into the same
file). Default: false. Incompatible with the I<snapshot> option.

=item snapshot

If this is true, then the copying from PostgreSQL takes place in
serialized mode (transaction isolation level serializable), which
should ensure consistency of relations between tables linked by
foreign key constraints. Currently, foreign keys are not created on
the SQLite side, however. Default: false. Incompatible with the
I<append> option.

=item cachedir

The current method for getting information about table structure in
PostgreSQL is somewhat slow, especially for databases with very many
tables. To offset this, table definitions are cached in a temporary
directory so that subsequent mirrorings of the same table will go
faster. The downside is, of course, that if the table structure
changes, the cache needs to be cleared manually. The cache directory
can be specified using this option; the default is
/tmp/sqlite_mirror_cache (with separate subdirectories for each user).

=item verbose

If this is true, a few messages will be output to stderr during the
mirroring process.

=back

=head1 TODO

=over 4

=item *

Support for foreign keys is missing.

=item *

The method used to read tables bigger than I<page_limit> needs to be
improved.

=item *

It would be nice to have a quick way of telling whether the cached
table definition of a specific table is still valid.

=item *

Tests.

=back

=head1 AUTHOR

Baldur Kristinsson (bk@mbl.is), 2004-2006.

 Copyright (c) 2006 Baldur Kristinsson. All rights reserved.
 This program is free software; you can redistribute it and/or
 modify it under the same terms as Perl itself.

=cut

